pkg/injector/patch.go

```diff
@@ -139,41 +135,67 @@ func addContainer(target, add []corev1.Container, basePath string) (patch []JSON
 			path += "/-"
 		}
 		patch = append(patch, JSONPatchOperation{
-			Op:    "add",
+			Op:    addOperation.String(),
 			Path:  path,
 			Value: value,
 		})
 	}
 	return patch
 }
 
-func updateAnnotation(target, add map[string]string, basePath string) (patch []JSONPatchOperation) {
+type operation string
+
+func (o operation) String() string {
+	return string(o)
+}
+
+const (
+	addOperation     operation = "add"
+	replaceOperation operation = "replace"
+)
+
+func getOpAndPath(target map[string]string, key, basePath string) (patchOperation operation, patchPath string) {
+	if target == nil {
+		// This is an addition
+		return addOperation, basePath
+	}
+
+	if target[key] != "" {
+		// This is a replacement
+		return replaceOperation, path.Join(basePath, escapeJSONPointerValue(key))
+	}
+
+	// Target exists, but the key is new.
+	return addOperation, basePath // path.Join(basePath, escapeJSONPointerValue(key))
+}
+
+func updateAnnotation(target, add map[string]string, basePath string) []JSONPatchOperation {
+	var patches []JSONPatchOperation
+
 	for key, value := range add {
-		if target == nil {
-			// First one will be a Create
-			target = map[string]string{}
-			patch = append(patch, JSONPatchOperation{
-				Op:   "add",
-				Path: basePath,
-				Value: map[string]string{
+		patchOperation, patchPath := getOpAndPath(target, key, basePath)
+
+		patchValue := map[operation]func() interface{}{
+
+			addOperation: func() interface{} {
+				return map[string]string{
 					key: value,
-				},
-			})
-		} else {
-			// Update
-			op := "add"
-			if target[key] != "" {
-				op = "replace"
-			}
-			patch = append(patch, JSONPatchOperation{
-				Op:    op,
-				Path:  path.Join(basePath, escapeJSONPointerValue(key)),
-				Value: value,
-			})
-		}
+				}
+			},
+
+			replaceOperation: func() interface{} {
+				return value
+			},
+		}[patchOperation]()
+
+		patches = append(patches, JSONPatchOperation{
+			Op:    patchOperation.String(),
+			Path:  patchPath,
+			Value: patchValue,
```

Comment:
I still don't understand what is wrong with the existing implementation, do you mind adding more details?

Check https://github.com/openservicemesh/osm/pull/1772#issuecomment-702252547

Reply:
I noticed this function was non deterministic.  Decided to rewrite it - but it was not correct.
To add determinism I reverted to the correct way you had implemented previously @shashankram, but added an iteration over sorted map keys.